spec/lib/test_executor_spec.rb

```diff
@@ -0,0 +1,91 @@
+require './app/lib/test_executor.rb'
+require 'helpers/rails_helper'
+require 'helpers/api_helper'
+
+RSpec.describe TestExecutor do
+  include ActionDispatch::TestProcess
+
+  let(:student) { create(:student) }
+  let(:teacher) { create(:teacher) }
+  let(:course) { create(:course) }
+  let(:assignment) { create(:assignment, course: course) }
+  let!(:problem) { create(:problem, assignment: assignment) }
+  let(:good_solution_file) { fixture_file_upload("#{fixture_path}/files/Solution.java") }
+  let(:submission_db_file) { create(:submission_db_file,
+                                    name: good_solution_file.original_filename,
+                                    contents: good_solution_file.read) }
+  let!(:submission) { create(:submission, user: student, problem: problem, db_file_id: submission_db_file.id) }
+
+  subject { described_class }
+
+  describe ".create_testing_image" do
+    # All expectations in one test, so we only actually create the image once.
+    it "creates a singleton Docker::Image for the Submission" do
+      expect(subject.images).to be_empty
+      expect(Docker::Image).to receive(:build_from_dir).once.and_call_original
+
+      image = subject.create_testing_image(submission)
+      subject.create_testing_image(submission)
+
+      expect(image).to be_instance_of(Docker::Image)
+      expect(subject.images[submission.language]).to eq(image)
+    end
+  end
+
+  context "when executing tests with an existing image" do
+    let(:consec_3) { create(:consec, n: 3) }
+    let(:consec_5) { create(:consec, n: 5) }
+    let(:test_consec_3) { create(:test, user: student, problem: problem, db_file_id: consec_3.id) }
+    let(:test_consec_5) { create(:test, user: student, problem: problem, db_file_id: consec_5.id) }
+    let(:bad_solution_file) { fixture_file_upload("#{fixture_path}/files/BadSolution.java") }
+    let(:bad_submission_db_file) { create(:submission_db_file,
+                                          name: bad_solution_file.original_filename,
+                                          contents: bad_solution_file.read) }
+    let!(:bad_submission) { create(:submission, user: student, problem: problem, db_file_id: bad_submission_db_file.id) }
+
+    before do
+      subject.create_testing_image(submission)
+
+      # solution_id needs to be set after both objects have been created
+      problem.solution_id = submission.id
```

Comment:
This line confused me and it convinces me that we should rename some of the variables.
```
submission -> solution
submission_db_file -> solution_db_file
good_solution_file -> solution_file
bad_solution_file -> bad_submission_file
```

A `solution` should by default be a perfect submission. A `bad_solution` shouldn't exist, it should just be `bad_submission`